### DISCUSSION TITLE
same time on two machines using horovod

### DISCUSSION THREAD

**Discussion by he2077** [horovod/horovod#2762] [Q&A]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)
2. Framework version:
3. Horovod version:
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version:
8. OS and version:
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
Hi,
I've been following horovod tutorial to transform a regular ssd model into a distributed one.
However, when tested with two machines, it has the same training time as one machine. So I presume some code I wrote for the horovod part is wrong. Any insights on which part is wrong? Thanks in advance!

`from data import *
from utils.augmentations import SSDAugmentation
from layers.modules import MultiBoxLoss
from ssd import build_ssd
import os
import sys
import time
import torch
from torch.autograd import Variable
import torch.nn as nn
import torch.optim as optim
import torch.backends.cudnn as cudnn
import torch.nn.init as init
import torch.utils.data as data
import numpy as np
import argparse
import horovod.torch as hvd
import tarfile
from ftplib import FTP

def str2bool(v):
    return v.lower() in ("yes", "true", "t", "1")

root_dir = '/usr/share/horovod'

parser = argparse.ArgumentParser(
    description='Single Shot MultiBox Detector Training With Pytorch')
train_set = parser.add_mutually_exclusive_group()
parser.add_argument('--dataset', default='VOC', choices=['VOC', 'COCO'],
                    type=str, help='VOC or COCO')
parser.add_argument('--dataset_root', default='VOCdevkit',
                    help='Dataset root directory path')
parser.add_argument('--basenet', default='vgg16_reducedfc.pth',
                    help='Pretrained base model')
parser.add_argument('--batch_size', default=32, type=int,
                    help='Batch size for training')
parser.add_argument('--resume', default=None, type=str,
                    help='Checkpoint state_dict file to resume training from')
parser.add_argument('--start_iter', default=0, type=int,
                    help='Resume training at this iter')
parser.add_argument('--num_workers', default=0, type=int,
                    help='Number of workers used in dataloading')
parser.add_argument('--cuda', default=True, type=str2bool,
                    help='Use CUDA to train model')
parser.add_argument('--lr', '--learning-rate', default=0.5e-3, type=float,
                    help='initial learning rate')
parser.add_argument('--momentum', default=0.9, type=float,
                    help='Momentum value for optim')
parser.add_argument('--weight_decay', default=5e-4, type=float,
                    help='Weight decay for SGD')
parser.add_argument('--gamma', default=0.1, type=float,
                    help='Gamma update for SGD')
parser.add_argument('--visdom', default=False, type=str2bool,
                    help='Use visdom for loss visualization')
parser.add_argument('--save_folder', default='weights/',
                    help='Directory for saving checkpoint models')
parser.add_argument('--seed', default=42,metavar='S',help='random seed')
parser.add_argument('--max_iter', default=5170,type=int)
parser.add_argument('--save_checkpoint_interval', default=5000, type=int)
parser.add_argument("--user_id",type=int,default=1000,help="user_id")
parser.add_argument("--task_id",type=int,default=1000,help="task_id")
args = parser.parse_args()

if torch.cuda.is_available():
    if args.cuda:
        torch.set_default_tensor_type('torch.cuda.FloatTensor')
    if not args.cuda:
        print("WARNING: It looks like you have a CUDA device, but aren't " +
              "using CUDA.\nRun with --cuda for optimal training speed.")
        torch.set_default_tensor_type('torch.FloatTensor')
else:
    torch.set_default_tensor_type('torch.FloatTensor')

if not os.path.exists(args.save_folder):
    os.mkdir(args.save_folder)

def train():
    if args.dataset == 'COCO':
        if args.dataset_root == VOC_ROOT:
            if not os.path.exists(COCO_ROOT):
                parser.error('Must specify dataset_root if specifying dataset')
            print("WARNING: Using default COCO dataset_root because " +
                  "--dataset_root was not specified.")
            args.dataset_root = COCO_ROOT
        cfg = coco
        dataset = COCODetection(root=args.dataset_root,
                                transform=SSDAugmentation(cfg['min_dim'],
                                                          MEANS))
    elif args.dataset == 'VOC':
        if args.dataset_root == COCO_ROOT:
            parser.error('Must specify dataset if specifying dataset_root')
        cfg = voc
        dataset = VOCDetection(root=args.dataset_root,
                               transform=SSDAugmentation(cfg['min_dim'],
                                                         MEANS))

    if args.visdom:
        import visdom
        viz = visdom.Visdom()

    ssd_net = build_ssd('train', cfg['min_dim'], cfg['num_classes'])
    net = ssd_net

    if args.cuda:
        net = torch.nn.DataParallel(ssd_net)
        cudnn.benchmark = True

    if args.resume:
        print('Resuming training, loading {}...'.format(args.resume))
        sys.stdout.flush()
        ssd_net.load_weights(args.resume)
    else:
        vgg_weights = torch.load(args.save_folder + args.basenet)
        print('Loading base network...')
        sys.stdout.flush()
        ssd_net.vgg.load_state_dict(vgg_weights)

    if args.cuda:
        net = net.cuda()

    if not args.resume:
        print('Initializing weights...')
        sys.stdout.flush()
        # initialize newly added layers' weights with xavier method
        ssd_net.extras.apply(weights_init)
        ssd_net.loc.apply(weights_init)
        ssd_net.conf.apply(weights_init)

    optimizer = optim.SGD(net.parameters(), lr=args.lr*hvd.size(), momentum=args.momentum,
                          weight_decay=args.weight_decay)
    hvd.broadcast_parameters(net.state_dict(),root_rank=0)
    hvd.broadcast_optimizer_state(optimizer,root_rank=0)
    optimizer = hvd.DistributedOptimizer(optimizer,named_parameters = net.named_parameters(),
            op=hvd.Average)
    criterion = MultiBoxLoss(cfg['num_classes'], 0.5, True, 0, True, 3, 0.5,
                             False, args.cuda)

    net.train()
    # loss counters
    loc_loss = 0
    conf_loss = 0
    epoch = 0
    print('Loading the dataset...')
    sys.stdout.flush()

    epoch_size = len(dataset) // args.batch_size
    print("the size is")
    print(len(dataset))
    print(epoch_size)
    print('Training SSD on:', dataset.name)
    sys.stdout.flush()
    print('Using the specified args:')
    sys.stdout.flush()
    print(args)
    sys.stdout.flush()

    step_index = 0

    if args.visdom:
        vis_title = 'SSD.PyTorch on ' + dataset.name
        vis_legend = ['Loc Loss', 'Conf Loss', 'Total Loss']
        iter_plot = create_vis_plot('Iteration', 'Loss', vis_title, vis_legend)
        epoch_plot = create_vis_plot('Epoch', 'Loss', vis_title, vis_legend)
    train_sampler = torch.utils.data.distributed.DistributedSampler(dataset,num_replicas=hvd.size(),
            rank=hvd.rank())
    data_loader = data.DataLoader(dataset, args.batch_size,
                                  sampler=train_sampler,
                                  num_workers=args.num_workers,
                                  collate_fn=detection_collate,
                                  pin_memory=True)
    # create batch iterator
    batch_iterator = iter(data_loader)
    for iteration in range(args.start_iter, args.max_iter):
        if args.visdom and iteration != 0 and (iteration % epoch_size == 0):
            update_vis_plot(epoch, loc_loss, conf_loss, epoch_plot, None,
                            'append', epoch_size)
            # reset epoch loss counters
            loc_loss = 0
            conf_loss = 0
            epoch += 1

        if iteration in cfg['lr_steps']:
            step_index += 1
            adjust_learning_rate(optimizer, args.gamma, step_index)

        # load train data
        try:
            images, targets = next(batch_iterator)
        except StopIteration as e:
            batch_iterator = iter(data_loader)
            images, targets = next(batch_iterator)

        if args.cuda:
            images = Variable(images.cuda())
            targets = [Variable(ann.cuda(), volatile=True) for ann in targets]
        else:
            images = Variable(images)
            targets = [Variable(ann, volatile=True) for ann in targets]
        # forward
        t0 = time.time()
        out = net(images)
        # backprop
        optimizer.zero_grad()
        loss_l, loss_c = criterion(out, targets)
        loss = loss_l + loss_c
        loss.backward()
        optimizer.step()
        t1 = time.time()
        loc_loss += loss_l.item()
        conf_loss += loss_c.item()

        if iteration % 10 == 0:
            print('timer: %.4f sec.' % (t1 - t0))
            sys.stdout.flush()
            print('iter ' + repr(iteration) + ' || Loss: %.4f ||' % (loss.item()), end=' ')
            sys.stdout.flush()

        if args.visdom:
            update_vis_plot(iteration, loss_l.data[0], loss_c.data[0],
                            iter_plot, epoch_plot, 'append')

        if iteration != 0 and iteration % args.save_checkpoint_interval == 0:
            print('Saving state, iter:', iteration)
            sys.stdout.flush()
            torch.save(ssd_net.state_dict(), 'weights/ssd300_COCO_' +
                       repr(iteration) + '.pth')
    if hvd.rank() == 0:
        torch.save(ssd_net.state_dict(),
                args.save_folder + '' + args.dataset + '.pth')
        ftp_flag = True
        while ftp_flag:
            try:
                ftp_flag = False
                ftp = FTP('172.18.29.81')
                ftp.login('ftper', 'admin')
                if args.user_id == 1000:
                    ftp.cwd('checkpoint')
                else:
                    ftp.cwd('user/%s/%s/result'%(args.user_id, args.task_id))
            except:
                print('trying to reconnect')
                sys.stdout.flush()
                ftp_flag = True
        bufsize = 1024
        output_file = ('ssd_iteration_%d.tar.gz'%(iteration+1))
        tar = tarfile.open(output_file, 'w')
        filepath = args.save_folder + '' + args.dataset + '.pth'
        tar.add(filepath)
        tar.close()
        fp = open(output_file, 'rb')
        ftp.storbinary('STOR ' + output_file.split('/')[-1], fp, bufsize)
        fp.close()
        ftp.quit()
        os.system('rm %s'%(filepath))

def adjust_learning_rate(optimizer, gamma, step):
    """Sets the learning rate to the initial LR decayed by 10 at every
        specified step
    # Adapted from PyTorch Imagenet example:
    # https://github.com/pytorch/examples/blob/master/imagenet/main.py
    """
    lr = args.lr * (gamma ** (step))
    for param_group in optimizer.param_groups:
        param_group['lr'] = lr

def xavier(param):
    init.xavier_uniform(param)

def weights_init(m):
    if isinstance(m, nn.Conv2d):
        xavier(m.weight.data)
        m.bias.data.zero_()

def create_vis_plot(_xlabel, _ylabel, _title, _legend):
    return viz.line(
        X=torch.zeros((1,)).cpu(),
        Y=torch.zeros((1, 3)).cpu(),
        opts=dict(
            xlabel=_xlabel,
            ylabel=_ylabel,
            title=_title,
            legend=_legend
        )
    )

def update_vis_plot(iteration, loc, conf, window1, window2, update_type,
                    epoch_size=1):
    viz.line(
        X=torch.ones((1, 3)).cpu() * iteration,
        Y=torch.Tensor([loc, conf, loc + conf]).unsqueeze(0).cpu() / epoch_size,
        win=window1,
        update=update_type
    )
    # initialize epoch plot on first iteration
    if iteration == 0:
        viz.line(
            X=torch.zeros((1, 3)).cpu(),
            Y=torch.Tensor([loc, conf, loc + conf]).unsqueeze(0).cpu(),
            win=window2,
            update=True
        )

if __name__ == '__main__':
    hvd.init()
    torch.manual_seed(args.seed)
    if args.cuda:
        torch.cuda.set_device(hvd.local_rank())
        torch.cuda.manual_seed(args.seed)
    train()
`

**Comment by he2077** [comment]
Another follow up question is about yolov3, where same thing happened, two machine same time as one machine, code is as following

from __future__ import division
import torch
print(torch.cuda.is_available())
from models import *
from utils.logger import *
from utils.utils import *
from utils.datasets import *
from utils.parse_config import *
#from test import evaluate

from terminaltables import AsciiTable

import warnings
warnings.filterwarnings('ignore')

import os
import sys
import time
import datetime
import argparse
from ftplib import FTP
import tarfile
import urllib.request

import torch
from torch.utils.data import DataLoader
from torchvision import datasets
from torchvision import transforms
from torch.autograd import Variable
import torch.optim as optim
import horovod.torch as hvd
print(torch.cuda.is_available())

from xml2model_pytorch import ModelConstructer, TrainSetting

root_dir = os.getcwd() + '/'

if __name__ == "__main__":
    hvd.init()
    print(torch.cuda.is_available())
    parser = argparse.ArgumentParser()
    parser.add_argument("--gradient_accumulations", type=int, default=2, help="number of gradient accums before step")
    parser.add_argument("--model_def", type=str, default=root_dir + "config/yolov3_cv_multi.cfg", help="path to model definition file")
    parser.add_argument("--data_config", type=str, default=root_dir + "config/0601.data", help="path to data config file")
    parser.add_argument("--pretrained_weights", type=str, default=root_dir + "yolov3_cv_multi_0601_cv_model.pth",help="if specified starts from checkpoint model")
    parser.add_argument("--checkpoints_dir", type=str, default=root_dir + "checkpoints", help="path to save weight file")

    parser.add_argument("--n_gpu", type=int, default=0, help="number of gpu threads to use during batch generation")
    parser.add_argument("--img_size", type=int, default=416, help="size of each image dimension")
    parser.add_argument("--checkpoint_interval", type=int, default=1, help="interval between saving model weights")                                  
    parser.add_argument("--evaluation_interval", type=int, default=1, help="interval evaluations on validation set")
    parser.add_argument("--compute_map", default=True, help="if True computes mAP every tenth batch")
    parser.add_argument("--multiscale_training", default=True, help="allow for multi-scale training")
    parser.add_argument("--backward-passes-per-step",type=float,default=1,help="number of backward iterations steps")
    parser.add_argument("--user_id",type=int,default=1000,help="user_id")
    parser.add_argument("--task_id",type=int,default=1000,help="task_id")
    opt = parser.parse_args()
    logger = Logger("logs")
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    torch.cuda.set_device(hvd.local_rank())
    os.makedirs("output", exist_ok=True)
    os.makedirs(opt.checkpoints_dir, exist_ok=True)

    # Get data configuration
    data_config = parse_data_config(opt.data_config)
    train_path = data_config["train"]
    valid_path = data_config["valid"]
    class_names = load_classes(data_config["names"])

    # Initiate model
    #model = Darknet(opt.model_def).to(device)
    #model.apply(weights_init_normal)

    model = ModelConstructer('tmp.xml').to(device)
    model.apply(weights_init_normal)

    run_epochs = TrainSetting.epochs
    batch_size = TrainSetting.batch_size 

    # Get dataloader
    dataset = ListDataset(train_path, augment=True, multiscale=opt.multiscale_training)
    train_sampler = torch.utils.data.DistributedSampler(dataset, num_replicas=hvd.size(),
            rank=hvd.rank())
    dataloader = torch.utils.data.DataLoader(
        dataset,
        batch_size=batch_size,
        sampler=train_sampler,
        num_workers=opt.n_gpu,
        collate_fn=dataset.collate_fn,
        pin_memory=True
    )

    lr_scaler = hvd.size()
    # 
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-5*lr_scaler, weight_decay=5e-4)
    hvd.broadcast_parameters(model.state_dict(),root_rank=0)
    hvd.broadcast_optimizer_state(optimizer,root_rank=0)
    optimizer = hvd.DistributedOptimizer(optimizer,named_parameters=model.named_parameters(),
            op=hvd.Average,backward_passes_per_step=2)
    metrics = [
        "grid_size",
        "loss",
        "x",
        "y",
        "w",
        "h",
        "conf",
        "cls",
        "cls_acc",
        "recall50",
        "recall75",
        "precision",
        "conf_obj",
        "conf_noobj",
    ]

    for epoch in range(run_epochs):
        model.train()
        train_sampler.set_epoch(epoch)
        start_time = time.time()
        for batch_i, (_, imgs, targets) in enumerate(dataloader):
            print(targets)
            batches_done = len(dataloader) * epoch + batch_i

            imgs = Variable(imgs.to(device))
            targets = Variable(targets.to(device), requires_grad=False)
            y = model(imgs, targets)
            loss = sum([y_i[1] for y_i in y])
            outputs = torch.cat([y_i[0] for y_i in y], 1)
            #loss, outputs = model(imgs, targets)
            loss = loss / 2
            loss.backward()

            if batches_done % opt.gradient_accumulations:
                # Accumulates gradient before each step
                optimizer.step()
                optimizer.zero_grad()

            # ----------------
            #   Log progress
            # ----------------
            if batch_i % 10 != 0:
                continue
            log_str = "\n---- [Epoch %d/%d, Batch %d/%d] ----\n" % (epoch, run_epochs, batch_i, len(dataloader))
            metric_table = [["Metrics", *[f"YOLO Layer {i}" for i in range(len(model.yolo_layers))]]]
            log_str += f"\nTotal loss {loss.item()}"
            # Determine approximate time left for epoch
            epoch_batches_left = len(dataloader) - (batch_i + 1)
            time_left = datetime.timedelta(seconds=epoch_batches_left * (time.time() - start_time) / (batch_i + 1))
            log_str += f"\n---- ETA {time_left}"
            print(log_str)

            model.seen += imgs.size(0)


        if epoch % opt.checkpoint_interval == 0 and hvd.rank() == 0:
            torch.save(model.state_dict(), f"{opt.checkpoints_dir}/yolov3_ckpt_%d.pth" % epoch)
            # torch.save(model, f"{opt.checkpoints_dir}/yolov3_ckpt_%d.pth" % epoch)
            ftp_flag = True
            while ftp_flag:
                try:
                    ftp_flag = False
                    ftp = FTP('172.18.29.81')
                    ftp.login('ftper', 'admin')
                    if opt.user_id == 1000:
                        ftp.cwd('checkpoint')
                    else:
                        ftp.cwd('user')
                        ftp.cwd(str(opt.user_id))
                        ftp.cwd(str(opt.task_id))
                        ftp.cwd('result')
                except:
                    print('trying to reconnect')
                    ftp_flag = True
            bufsize = 1024
            output_file = ('%s/yolov3_ckpt_%d.tar.gz'%(opt.checkpoints_dir, epoch)) 
            tar = tarfile.open(output_file, 'w')
            tar.add('%s/yolov3_ckpt_%d.pth'%(opt.checkpoints_dir, epoch))
            tar.close()
            fp = open(output_file, 'rb')
            ftp.storbinary('STOR ' + output_file.split('/')[-1], fp, bufsize)
            fp.close()
            ftp.quit()
            print('Training Done')

**Comment by tgaddair** [comment]
Hey @erichu2000, I suspect the issue is here:

```
for iteration in range(args.start_iter, args.max_iter):
```

Looks like you're not scaling the number of steps proportional to the number of workers.

In the second example, it's less clear what's happening.  I would try running a script that prints `hvd.size()` and make sure that it's equal to the number of workers you're attempting to train with.  I would also suggest running the `pytorch_mnist.py` example script for Horovod to make sure that your environment is setup correctly.  The number of steps per epoch should decrease with the number of workers in both the example script and in your second script.

**Comment by densechen** [comment]
Hi, @tgaddair, I also meet the same question. And my code is modified from the pytorch_mnist.py.

```
"""Train CIFAR with PyTorch

The hyperparameter is token from AdaBelief.
"""
import argparse
import os
import random
import time
import horovod.torch as hvd
import torch as th
import torch.nn as nn
import torch.optim as optim
import torch.utils.data.distributed
import torchvision
import torchvision.transforms as transforms
import models
parser = argparse.ArgumentParser(description='PyTorch CIFAR Training')
parser.add_argument('--resume', '-r', action='store_true',
                    help='resume from checkpoint')
parser.add_argument('--bs', type=int, default=128, help='batch size')
parser.add_argument('--seed', type=int, default=0,
                    help='random seed (default: 0')
parser.add_argument('--fp16-allreduce', action='store_true', default=False,
                    help='use fp16 compression during allreduce')
parser.add_argument('--use-adasum', action='store_true', default=False,
                    help='use adasum algorithm to do reduction')
parser.add_argument('--gradient-predivide-factor', type=float, default=1.0,
                    help='apply gradient predivide factor in optimizer (default: 1.0)')
parser.add_argument('--batches-per-allreduce', type=int, default=1,
                    help='number of batches processed locally before '
                         'executing allreduce across workers; it multiplies '
                         'total batch size.')
args = parser.parse_args()
def build_dataset():
    train_sampler = torch.utils.data.distributed.DistributedSampler(
        trainset, num_replicas=hvd.size(), rank=hvd.rank(), shuffle=True)
    train_loader = torch.utils.data.DataLoader(
        trainset, batch_size=args.bs, shuffle=train_sampler is None,
        num_workers=4, sampler=train_sampler)
    test_sampler = torch.utils.data.distributed.DistributedSampler(
        testset, num_replicas=hvd.size(), rank=hvd.rank(), shuffle=False)
    test_loader = torch.utils.data.DataLoader(
        testset, batch_size=args.bs, shuffle=test_sampler is None,
        num_workers=4, sampler=test_sampler)
    return train_loader, test_loader
def train(net, epoch, data_loader, inner_optimizer, outter_optimizer, criterion):
    net.train()
    train_loss = 0
    correct = 0
    total = 0
    # do not forget
    data_loader.sampler.set_epoch(epoch)
    for _, (inputs, targets) in enumerate(data_loader):
        inputs, targets = inputs.cuda(), targets.cuda()
        if inner_optimizer is not None:
            inner_optimizer.zero_grad()
        outter_optimizer.zero_grad()
        outputs = net(inputs)
        loss = criterion(outputs, targets)
        loss.backward()
        if inner_optimizer is not None:
            inner_optimizer.step()
        byzantine.TrojanHorse.random()
        outter_optimizer.step()
        train_loss += loss.item()
        _, predicted = outputs.max(1)
        total += targets.size(0)
        correct += predicted.eq(targets).sum()
    accuracy = 100. * correct / total
    # Horovod: average metric values across workers.
    accuracy = metric_average(accuracy, 'accuracy')
    return accuracy
def test(model, test_loader):
    model.eval()
    correct = 0
    total = 0
    for data, target in test_loader:
        data, target = data.cuda(), target.cuda()
        output = model(data)
        # get the index of the max log-probability
        pred = output.data.max(1, keepdim=True)[1]
        total += target.size(0)
        correct += pred.eq(target.data.view_as(pred)).cpu().float().sum()
    # Horovod: use test_sampler to determine the number of examples in
    # this worker's partition.
    accuracy = 100. * correct / total
    # Horovod: average metric values across workers.
    test_accuracy = metric_average(accuracy, 'avg_accuracy')
    return test_accuracy
if __name__ == '__main__':
    args.cuda = th.cuda.is_available()
    # Horovod: initialize library.
    hvd.init()
    th.manual_seed(args.seed)
    random.seed(args.seed)
    if args.cuda:
        # Horovod: pin GPU to local rank
        th.cuda.set_device(hvd.local_rank())
        th.cuda.manual_seed(args.seed)
    # Horovod: limit # of CPU threads to be used per worker.
    th.set_num_threads(4)
    train_loader, test_loader = build_dataset()
    net = build_model()
    criterion = nn.CrossEntropyLoss()
    outter_optimizer = create_optimizer(net.parameters())
    # Horovod: broadcast parameters & optimizer state.
    hvd.broadcast_parameters(net.state_dict(), root_rank=0)
    hvd.broadcast_optimizer_state(outter_optimizer, root_rank=0)
    # Horovod: (optional) compression algorithm.
    compression = hvd.Compression.fp16 if args.fp16_allreduce else hvd.Compression.none
    # Horovod: wrap optimizer with DistributedOptimizer.
    outter_optimizer = hvd.DistributedOptimizer(outter_optimizer,
                                                named_parameters=net.named_parameters(),
                                                compression=compression,
                                                op=hvd.Adasum if args.use_adasum else hvd.Average,
                                                gradient_predivide_factor=args.gradient_predivide_factor)
    for epoch in range(1, args.epoch+1):
        start = time.time()
        train_acc = train(net, epoch, train_loader,
                          inner_optimizer, outter_optimizer, criterion)
        test_acc = test(net, test_loader)
        end = time.time()
```

**Comment by tgaddair** [comment]
Hey @densechen, looks like you're using the `DistributedSampler`, so the splitting of the data should work correctly. Can you try running a script that just prints out `hvd.rank()` and `hvd.size()` to make sure all the workers are initialized correctly?

**Comment by densechen** [comment]
> Hi, @tgaddair, I also meet the same question. And my code is modified from the pytorch_mnist.py.
> 
> ```
> """Train CIFAR with PyTorch
> 
> The hyperparameter is token from AdaBelief.
> """
> import argparse
> import os
> import random
> import time
> import horovod.torch as hvd
> import torch as th
> import torch.nn as nn
> import torch.optim as optim
> import torch.utils.data.distributed
> import torchvision
> import torchvision.transforms as transforms
> import models
> parser = argparse.ArgumentParser(description='PyTorch CIFAR Training')
> parser.add_argument('--resume', '-r', action='store_true',
>                     help='resume from checkpoint')
> parser.add_argument('--bs', type=int, default=128, help='batch size')
> parser.add_argument('--seed', type=int, default=0,
>                     help='random seed (default: 0')
> parser.add_argument('--fp16-allreduce', action='store_true', default=False,
>                     help='use fp16 compression during allreduce')
> parser.add_argument('--use-adasum', action='store_true', default=False,
>                     help='use adasum algorithm to do reduction')
> parser.add_argument('--gradient-predivide-factor', type=float, default=1.0,
>                     help='apply gradient predivide factor in optimizer (default: 1.0)')
> parser.add_argument('--batches-per-allreduce', type=int, default=1,
>                     help='number of batches processed locally before '
>                          'executing allreduce across workers; it multiplies '
>                          'total batch size.')
> args = parser.parse_args()
> def build_dataset():
>     train_sampler = torch.utils.data.distributed.DistributedSampler(
>         trainset, num_replicas=hvd.size(), rank=hvd.rank(), shuffle=True)
>     train_loader = torch.utils.data.DataLoader(
>         trainset, batch_size=args.bs, shuffle=train_sampler is None,
>         num_workers=4, sampler=train_sampler)
>     test_sampler = torch.utils.data.distributed.DistributedSampler(
>         testset, num_replicas=hvd.size(), rank=hvd.rank(), shuffle=False)
>     test_loader = torch.utils.data.DataLoader(
>         testset, batch_size=args.bs, shuffle=test_sampler is None,
>         num_workers=4, sampler=test_sampler)
>     return train_loader, test_loader
> def train(net, epoch, data_loader, inner_optimizer, outter_optimizer, criterion):
>     net.train()
>     train_loss = 0
>     correct = 0
>     total = 0
>     # do not forget
>     data_loader.sampler.set_epoch(epoch)
>     for _, (inputs, targets) in enumerate(data_loader):
>         inputs, targets = inputs.cuda(), targets.cuda()
>         if inner_optimizer is not None:
>             inner_optimizer.zero_grad()
>         outter_optimizer.zero_grad()
>         outputs = net(inputs)
>         loss = criterion(outputs, targets)
>         loss.backward()
>         if inner_optimizer is not None:
>             inner_optimizer.step()
>         byzantine.TrojanHorse.random()
>         outter_optimizer.step()
>         train_loss += loss.item()
>         _, predicted = outputs.max(1)
>         total += targets.size(0)
>         correct += predicted.eq(targets).sum()
>     accuracy = 100. * correct / total
>     # Horovod: average metric values across workers.
>     accuracy = metric_average(accuracy, 'accuracy')
>     return accuracy
> def test(model, test_loader):
>     model.eval()
>     correct = 0
>     total = 0
>     for data, target in test_loader:
>         data, target = data.cuda(), target.cuda()
>         output = model(data)
>         # get the index of the max log-probability
>         pred = output.data.max(1, keepdim=True)[1]
>         total += target.size(0)
>         correct += pred.eq(target.data.view_as(pred)).cpu().float().sum()
>     # Horovod: use test_sampler to determine the number of examples in
>     # this worker's partition.
>     accuracy = 100. * correct / total
>     # Horovod: average metric values across workers.
>     test_accuracy = metric_average(accuracy, 'avg_accuracy')
>     return test_accuracy
> if __name__ == '__main__':
>     args.cuda = th.cuda.is_available()
>     # Horovod: initialize library.
>     hvd.init()
>     th.manual_seed(args.seed)
>     random.seed(args.seed)
>     if args.cuda:
>         # Horovod: pin GPU to local rank
>         th.cuda.set_device(hvd.local_rank())
>         th.cuda.manual_seed(args.seed)
>     # Horovod: limit # of CPU threads to be used per worker.
>     th.set_num_threads(4)
>     train_loader, test_loader = build_dataset()
>     net = build_model()
>     criterion = nn.CrossEntropyLoss()
>     outter_optimizer = create_optimizer(net.parameters())
>     # Horovod: broadcast parameters & optimizer state.
>     hvd.broadcast_parameters(net.state_dict(), root_rank=0)
>     hvd.broadcast_optimizer_state(outter_optimizer, root_rank=0)
>     # Horovod: (optional) compression algorithm.
>     compression = hvd.Compression.fp16 if args.fp16_allreduce else hvd.Compression.none
>     # Horovod: wrap optimizer with DistributedOptimizer.
>     outter_optimizer = hvd.DistributedOptimizer(outter_optimizer,
>                                                 named_parameters=net.named_parameters(),
>                                                 compression=compression,
>                                                 op=hvd.Adasum if args.use_adasum else hvd.Average,
>                                                 gradient_predivide_factor=args.gradient_predivide_factor)
>     for epoch in range(1, args.epoch+1):
>         start = time.time()
>         train_acc = train(net, epoch, train_loader,
>                           inner_optimizer, outter_optimizer, criterion)
>         test_acc = test(net, test_loader)
>         end = time.time()
> ```

Solved in https://github.com/horovod/horovod/issues/2686.
